--- a/drivers/net/phy/at803x.c
+++ b/drivers/net/phy/at803x.c
@@ -255,6 +256,12 @@ static int at803x_probe(struct phy_devic
 	    phydev->drv->phy_id != ATH8032_PHY_ID)
 		goto does_not_require_reset_workaround;
 
+	pdata = dev_get_platdata(&phydev->dev);
+	if (pdata && pdata->has_reset_gpio) {
+		devm_gpio_request(dev, pdata->reset_gpio, "reset");
+		gpio_direction_output(pdata->reset_gpio, 1);
+	}
+
 	gpiod_reset = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_LOW);
 	if (IS_ERR(gpiod_reset))
 		return PTR_ERR(gpiod_reset);
